File: RadarSDK/RadarLog.m
//
//  RadarLog.m
//  RadarSDK
//
//  Copyright © 2021 Radar Labs, Inc. All rights reserved.
//

#import "RadarLog.h"
#import "RadarUtils.h"

@implementation RadarLog

- (instancetype _Nullable)initWithLevel:(RadarLogLevel)level type:(RadarLogType)type message:(NSString *_Nullable)message {
    self = [super init];
    if (self) {
        _level = level;
        _type = type;
        _message = message;
        _createdAt = [NSDate new];
    }
    return self;
}

+ (NSString *)stringForLogLevel:(RadarLogLevel)level {
    NSString *str;
    switch (level) {
    case RadarLogLevelNone:
        str = @"none";
        break;
    case RadarLogLevelError:
        str = @"error";
        break;
    case RadarLogLevelWarning:
        str = @"warning";
        break;
    case RadarLogLevelInfo:
        str = @"info";
        break;
    case RadarLogLevelDebug:
        str = @"debug";
        break;
    }
    return str;
}

+ (RadarLogLevel) levelFromString:(NSString *_Nonnull) string {
    if ([string isEqualToString:@"none"]) {
        return RadarLogLevelNone;
    } else if ([string isEqualToString:@"error"]) {
        return RadarLogLevelError;
    } else if ([string isEqualToString:@"warning"]) {
        return RadarLogLevelWarning;
    } else if ([string isEqualToString:@"info"]) {
        return RadarLogLevelInfo;
    } else if ([string isEqualToString:@"debug"]) {
        return RadarLogLevelDebug;
    } else {
        return RadarLogLevelInfo;
    }
}

+ (NSString *)stringForLogType:(RadarLogType)type {
    NSString *str;
    switch (type) {
    case RadarLogTypeNone:
        str = @"NONE";
        break;
    case RadarLogTypeSDKCall:
        str = @"SDK_CALL";
        break;
    case RadarLogTypeSDKError:
        str = @"SDK_ERROR";
        break;
    case RadarLogTypeSDKException:
        str = @"SDK_EXCEPTION";
        break;
    case RadarLogTypeAppLifecycleEvent:
        str = @"APP_LIFECYCLE_EVENT";
        break;
    case RadarLogTypePermissionEvent:
        str = @"PERMISSION_EVENT";
        break;
    }
    return str;
}

- (NSDictionary *)dictionaryValue {
    NSMutableDictionary *dict = [NSMutableDictionary new];
    dict[@"level"] = [RadarLog stringForLogLevel:self.level];
    dict[@"message"] = self.message;
    if (self.type) {
        dict[@"type"] = [RadarLog stringForLogType:self.type];
    }
    NSString *createdAtString = [RadarUtils.isoDateFormatter stringFromDate:self.createdAt];
    dict[@"createdAt"] = createdAtString;

    return dict;
}

+ (NSArray<NSDictionary *> *)arrayForLogs:(NSArray<RadarLog *> *)logs {
    if (!logs) {
        return nil;
    }

    NSMutableArray *arr = [[NSMutableArray alloc] initWithCapacity:logs.count];
    for (RadarLog *log in logs) {
        NSDictionary *dict = [log dictionaryValue];
        [arr addObject:dict];
    }
    return arr;
}

#pragma mark - NSCoding

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super init];
    if (self) {
        _level = [coder decodeIntegerForKey:@"level"];
        _type = [coder decodeIntegerForKey:@"type"];
        _message = [coder decodeObjectForKey:@"message"];
        _createdAt = [coder decodeObjectForKey:@"createdAt"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeInteger:_level forKey:@"level"];
    [coder encodeInteger:_type forKey:@"type"];
    [coder encodeObject:_message forKey:@"message"];
    [coder encodeObject:_createdAt forKey:@"createdAt"];
}

+ (BOOL)supportsSecureCoding {
    return YES;
}

@end

File: RadarSDK/RadarLog.swift
//
//  RadarLog.swift
//  RadarSDK
//
//  Created by ShiCheng Lu on 3/27/26.
//  Copyright © 2026 Radar Labs, Inc. All rights reserved.
//

import Foundation
import OSLog

extension RadarLogLevel: Codable {
    public func toString() -> String {
        switch self {
        case .none:
            return "none"
        case .error:
            return "error"
        case .warning:
            return "warning"
        case .info:
            return"info"
        case .debug:
            return "debug"
        @unknown default:
            return "none"
        }
    }
    
    public static func from(string: String) -> RadarLogLevel {
        switch string {
        case "none":
            return .none
        case "error":
            return .error
        case "warning":
            return .warning
        case "info":
            return .info
        case "debug":
            return .debug
        default:
            return .none
        }
    }
    
    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.toString())
    }
    
    public init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        self = RadarLogLevel.from(string: value)
    }
}

extension RadarLogType: Codable {
    public func toString() -> String {
        switch self {
        case .none:
            return "NONE"
        case .sdkCall:
            return "SDK_CALL"
        case .sdkError:
            return "SDK_ERROR"
        case .sdkException:
            return "SDK_EXCEPTION"
        case .appLifecycleEvent:
            return "APP_LIFECYCLE_EVENT"
        case .permissionEvent:
            return "PERMISSION_EVENT"
        @unknown default:
            return "NONE"
        }
    }
    
    public static func from(string: String) -> RadarLogType {
        switch string {
        case "NONE":
            return .none
        case "SDK_CALL":
            return .sdkCall
        case "SDK_ERROR":
            return .sdkError
        case "SDK_EXCEPTION":
            return .sdkException
        case "APP_LIFECYCLE_EVENT":
            return .appLifecycleEvent
        case "PERMISSION_EVENT":
            return .permissionEvent
        default:
            return .none
        }
    }
    
    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(self.toString())
    }
    
    public init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        self = RadarLogType.from(string: value)
    }
}

struct RadarLog: Codable, CustomStringConvertible {
    static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return formatter
    }()
    
    let level: RadarLogLevel
    let message: String
    let type: RadarLogType
    let createdAt: Date
    let includeDate: Bool
    let battery: Float?
    
    var description: String {
        var str = message
        if includeDate {
            let dateString = RadarLog.dateFormatter.string(from: createdAt)
            str += " | at \(dateString)"
        }
        if let battery {
            let batteryString = String(format: "%.2f", battery)
            str += " | with \(batteryString) battery"
        }
        return str
    }
    
    var osLogType: OSLogType {
        switch level {
        case .none:
            return .default
        case .error:
            return .error
        case .warning:
            return .error
        case .info:
            return .info
        case .debug:
            return .debug
        @unknown default:
            return .default
        }
    }
    
    func dict() -> [String: Any] {
        return [
            "message": description,
            "level": level.toString(),
            "type": type.toString(),
            "createdAt": RadarUtils.isoDateFormatter.string(from: createdAt),
        ]
    }
}
